src/main/java/org/spongepowered/api/command/args/CommandContext.java

```diff
@@ -170,4 +226,45 @@ public boolean hasAny(String key) {
     public boolean hasAny(Text key) {
         return hasAny(ArgUtils.textToArgKey(key));
     }
+
+    /**
+     * Gets a snapshot of the data inside this context to allow it to be
+     * restored later.
+     *
+     * <p>This is only guaranteed to create a <em>shallow copy</em> of the
+     * backing store. If any value is mutable, any changes to that value
+     * will be reflected in this snapshot. It is therefore not recommended
+     * that you keep this snapshot around for longer than is necessary.</p>
+     *
+     * @return The {@link Snapshot} containing the current state of the
+     *      {@link CommandContext}
+     */
+    public Snapshot getSnapshot() {
+        return new Snapshot(this.parsedArgs);
+    }
+
+    /**
+     * Resets a {@link CommandContext} to a previous state using a previously
+     * created {@link Snapshot}.
+     *
+     * @param snapshot The {@link Snapshot} to restore this context with
+     */
+    public void applySnapshot(Snapshot snapshot) {
+        this.parsedArgs.clear();
+        this.parsedArgs.putAll(snapshot.args);
+    }
+
+    /**
+     * A snapshot of a {@link CommandContext}. This object does not contain any
+     * public API methods, a snapshot should be considered a black box.
+     */
+    public class Snapshot {
```

Comment:
Maybe make it final?

Reply:
Agreed, should be final if you're not planning on allowing it to be extended.

Reply:
Yep, that will be done before I merge